File: packages/core/src/extract/extractor.ts
import type { Document, Element, ChildNode } from 'domhandler';
import { isTag } from 'domhandler';
import type { ExtractOptions } from '../types.js';
import { DEFAULT_STRIP_TAGS, DEFAULT_STRIP_ROLES, DEFAULT_STRIP_CLASSES, DEFAULT_STRIP_IDS } from './selectors.js';

/**
 * Remove non-content elements from a parsed HTML document in-place.
 *
 * Strips elements matching the built-in selectors (nav, ads, sidebars, etc.)
 * plus any additional patterns provided via {@link ExtractOptions}.
 *
 * @param document - The parsed DOM tree to prune.
 * @param options - Additional selectors and overrides.
 */
export function extractContent(document: Document, options?: ExtractOptions): void {
    const stripTags = new Set([...DEFAULT_STRIP_TAGS, ...(options?.stripTags ?? [])]);

    if (options?.keepHeader) stripTags.delete('header');
    if (options?.keepFooter) stripTags.delete('footer');
    if (options?.keepNav) stripTags.delete('nav');

    const stripRoles = new Set([...DEFAULT_STRIP_ROLES, ...(options?.stripRoles ?? [])]);

    const stripClasses = [...DEFAULT_STRIP_CLASSES, ...(options?.stripClasses ?? [])];

    const stripIds = [...DEFAULT_STRIP_IDS, ...(options?.stripIds ?? [])];

    pruneTree(document, stripTags, stripRoles, stripClasses, stripIds);
}

/**
 * Recursively walks the DOM tree and removes elements that match any
 * of the strip criteria. Removal is done in-place by splicing the
 * parent's `children` array.
 *
 * @param node - The parent node whose children will be inspected.
 * @param stripTags - Tag names to remove (e.g. `nav`, `footer`).
 * @param stripRoles - ARIA roles to remove (e.g. `navigation`, `banner`).
 * @param stripClasses - CSS class substrings or regex patterns to match for removal.
 * @param stripIds - Element ID substrings or regex patterns to match for removal.
 */
function pruneTree(
    node: Document | Element,
    stripTags: Set<string>,
    stripRoles: Set<string>,
    stripClasses: (string | RegExp)[],
    stripIds: (string | RegExp)[]
): void {
    const toRemove: ChildNode[] = [];

    for (const child of node.children) {
        if (!isTag(child)) continue;

        const el = child;

        if (shouldStrip(el, stripTags, stripRoles, stripClasses, stripIds)) {
            toRemove.push(child);
            continue;
        }

        pruneTree(el, stripTags, stripRoles, stripClasses, stripIds);
    }

    for (const child of toRemove) {
        const idx = node.children.indexOf(child);
        if (idx !== -1) {
            node.children.splice(idx, 1);
            child.parent = null;
        }
    }
}

/**
 * Determines whether an element should be stripped from the DOM tree.
 *
 * Checks against tag name, ARIA role, CSS class, and element ID in that order.
 *
 * @param el - The element to evaluate.
 * @param stripTags - Tag names to match.
 * @param stripRoles - ARIA roles to match.
 * @param stripClasses - CSS class substrings or regex patterns to match.
 * @param stripIds - Element ID substrings or regex patterns to match.
 * @returns `true` if the element matches any strip criterion.
 */
function shouldStrip(
    el: Element,
    stripTags: Set<string>,
    stripRoles: Set<string>,
    stripClasses: (string | RegExp)[],
    stripIds: (string | RegExp)[]
): boolean {
    if (stripTags.has(el.name)) return true;

    const role = el.attribs.role;
    if (role && stripRoles.has(role)) return true;

    const className = el.attribs.class;
    if (className && matchesAny(stripBrackets(className), stripClasses)) return true;

    const id = el.attribs.id;
    if (id && matchesAny(id, stripIds)) return true;

    return false;
}

/**
 * Remove Tailwind CSS arbitrary-value brackets from a class string so that
 * CSS custom property names (e.g. `[--fd-sidebar-width:268px]`) don't
 * trigger semantic class-name patterns like `/sidebar/`.
 */
function stripBrackets(value: string): string {
    return value.replaceAll(/\[[^\]]*\]/g, '');
}

/**
 * Tests a string against a list of patterns.
 *
 * @param value - The string to test (e.g. a class name or element ID).
 * @param patterns - String substrings (checked with `includes`) or `RegExp` instances.
 * @returns `true` if `value` matches at least one pattern.
 */
function matchesAny(value: string, patterns: (string | RegExp)[]): boolean {
    for (const pattern of patterns) {
        if ((typeof pattern === 'string' && value.includes(pattern)) || (pattern instanceof RegExp && pattern.test(value))) return true;
    }
    return false;
}

File: python/src/markdown_for_agents/extract/extractor.py
from __future__ import annotations

import re
from typing import TYPE_CHECKING

from ..dom.nodes import Document, Element, is_tag

if TYPE_CHECKING:
    from .._types import ExtractOptions
from .selectors import DEFAULT_STRIP_CLASSES, DEFAULT_STRIP_IDS, DEFAULT_STRIP_ROLES, DEFAULT_STRIP_TAGS


def extract_content(document: Document, options: ExtractOptions | None = None) -> None:
    """Remove non-content elements from a parsed HTML document in-place."""
    strip_tags = set(DEFAULT_STRIP_TAGS)
    if options is not None:
        strip_tags.update(options.strip_tags)
        if options.keep_header:
            strip_tags.discard("header")
        if options.keep_footer:
            strip_tags.discard("footer")
        if options.keep_nav:
            strip_tags.discard("nav")

    strip_roles = set(DEFAULT_STRIP_ROLES)
    if options is not None:
        strip_roles.update(options.strip_roles)

    strip_classes: list[str | re.Pattern[str]] = list(DEFAULT_STRIP_CLASSES)
    if options is not None:
        strip_classes.extend(options.strip_classes)

    strip_ids: list[str | re.Pattern[str]] = list(DEFAULT_STRIP_IDS)
    if options is not None:
        strip_ids.extend(options.strip_ids)

    _prune_tree(document, strip_tags, strip_roles, strip_classes, strip_ids)


def _prune_tree(
    node: Document | Element,
    strip_tags: set[str],
    strip_roles: set[str],
    strip_classes: list[str | re.Pattern[str]],
    strip_ids: list[str | re.Pattern[str]],
) -> None:
    """Recursively walk the DOM tree and remove elements matching strip criteria."""
    to_remove: list[Element] = []

    for child in node.children:
        if not is_tag(child):
            continue
        assert isinstance(child, Element)

        if _should_strip(child, strip_tags, strip_roles, strip_classes, strip_ids):
            to_remove.append(child)
            continue

        _prune_tree(child, strip_tags, strip_roles, strip_classes, strip_ids)

    for child in to_remove:
        if child in node.children:
            node.children.remove(child)
            child.parent = None


def _should_strip(
    el: Element,
    strip_tags: set[str],
    strip_roles: set[str],
    strip_classes: list[str | re.Pattern[str]],
    strip_ids: list[str | re.Pattern[str]],
) -> bool:
    """Determine whether an element should be stripped from the DOM tree."""
    if el.name in strip_tags:
        return True

    role = el.attribs.get("role", "")
    if role and role in strip_roles:
        return True

    class_name = el.attribs.get("class", "")
    if class_name and _matches_any(class_name, strip_classes):
        return True

    el_id = el.attribs.get("id", "")
    return bool(el_id and _matches_any(el_id, strip_ids))


def _matches_any(value: str, patterns: list[str | re.Pattern[str]]) -> bool:
    """Test a string against a list of patterns."""
    for pattern in patterns:
        if isinstance(pattern, str) and pattern in value or isinstance(pattern, re.Pattern) and pattern.search(value):
            return True
    return False
